src/components/ActionButton/StatefulActionButton.ts

```diff
@@ -0,0 +1,167 @@
+    /**
+     * An array containing all the state transitions allowed on the StatefulActionButton instance.
+     * If `switchTo` is called with an illegal transitions , a warning will be emitted
+     * and the transition will not occur.
+     */
+    allowedTransitions?: IStatefulActionButtonTransition[];
+    /**
+     * The initial state to be used by the StatefulActionButton instance.
+     * If nullish or missing from the states array, a warning will be emitted,
+     * no ActionButton will be constructed and the element will be hidden
+     */
+    initialState: StatefulActionButtonState;
+}
+
+/**
+ * An action button able to handle multiple states and their transitions.
+ */
+export class StatefulActionButton {
+    static ID = 'StatefulActionButton';
+    private currentState: StatefulActionButtonState;
+    private innerActionButton: ActionButton;
+
+    constructor(public element: HTMLElement, public options: IStatefulActionButtonOptions, public bindings?: IResultsComponentBindings) {
+        if (!this.areOptionsValid()) {
+            return;
+        }
+        this.currentState = this.options.initialState;
+        this.currentState.onStateEntry?.apply(this);
+        this.innerActionButton = new ActionButton(element, { ...this.options.initialState, click: this.handleClick.bind(this) }, bindings);
+    }
+
+    /**
+     * Switch the state of the instance if the state and the transition between the current and new state are allowed.
+     * @param state a state to try to switch to
+     */
+    public switchTo(state: StatefulActionButtonState) {
+        if (this.options.states.indexOf(state) === -1) {
+            console.warn(
+                `State '${state.name}' does not exist on this StatefulActionButton\nEnsure to use the object references used at the instantiation.`
+            );
+            return;
+        }
+        if (!this.isTransitionAllowed(state)) {
+            console.warn(
+                `Transition from State '${this.currentState.name}' to State '${state.name}' is not allowed on this StatefulActionButton.\nEnsure to use the object references used at the instantiation.`
+            );
+            return;
+        }
+        this.currentState.onStateExit?.apply(this);
+        state.onStateEntry?.apply(this);
+        this.innerActionButton.updateIcon(state.icon);
+        this.innerActionButton.updateTooltip(state.tooltip);
+        this.currentState = state;
+    }
+
+    /**
+     * Return the current state of the instance.
+     */
+    public getCurrentState() {
+        return this.currentState;
+    }
+
+    /**
+     * Check if the options given to the constructor are valid.
+     * If not, it will also display the appropriate warning.
+     */
+    private areOptionsValid(): boolean {
+        if (!this.options.states?.length) {
+            console.warn('The stateful action button cannot render if no states are defined.');
```

Comment:
One thing I considered was the use of `console.group` and transform the guard to checks, to ensure every check is run and provide a 'full' diagnostic.
What disappointed me is that `console.group` is stuck to the info level :(

I understand where this coming from, but if we were to display the warning `'Cannot render the stateful action button because options are invalid.'` it should be done before the details (e.g. no states are defined) I think.

I'll play with this idea for a bit, but if we're to do that, the smaller checks would make sense because areOptionsValid would just play its role of boolean determination and won't do the logging anymore, meaning it would look like a 
`return condA | condB | condC |...`